pkg/server/endpoints/node/handler.go

```diff
@@ -797,15 +821,94 @@ func (h *Handler) signCSRs(ctx context.Context,
 	return svids, nil
 }
 
+func (h *Handler) signCSRs(ctx context.Context,
+	peerCert *x509.Certificate, csrs map[string][]byte, regEntries []*common.RegistrationEntry) (
+	svids map[string]*node.X509SVID, err error) {
+
+	callerID, err := getSpiffeIDFromCert(peerCert)
+	if err != nil {
+		return nil, err
+	}
+
+	//convert registration entries into a map for easy lookup
+	regEntriesMap := make(map[string]*common.RegistrationEntry)
+	for _, entry := range regEntries {
+		regEntriesMap[entry.EntryId] = entry
+	}
+
+	ds := h.c.Catalog.GetDataStore()
+	svids = make(map[string]*node.X509SVID)
+	//iterate the CSRs and sign them
+	for entryID, csr := range csrs {
```

Comment:
As far as I can tell, the logic inside this loop is identical to that of signCSRsLegacy. In that case, does it make sense to extract this logic into a function signCSR? Then both signCSRsLegacy and signCSRs can loop over it independently and build the list types they need

Reply:
Hmm after I did this change, I realized that the function `signCSR(...)` would have to receive too many values... and perhaps it makes this logic more confusing.

Maybe it is not too bad to have CSRs signs logic for legacy and new agents split. In that way, future changes (not affecting the legacy CSRs sign) could be easier to do.

I have no problem to change it anyways if you think that it is better in the other way :slightly_smiling_face: